File: problems/912. Sort an Array/Solution2.cpp
// 快速排序-双边扫描
class Solution {
private:
    void quickSort(vector<int>& nums, int l, int u) {
        if (l >= u) return;

        // swap(nums[l], nums[l+(u-l)/2]);
        swap(nums[l], nums[rand()%(u-l+1)+l]);
        
        int t = nums[l], i = l, j = u + 1;
        while (i < j) {
            do ++i; while (i <= u && nums[i] < t);
            do --j; while (nums[j] > t);
            if (i >= j) break;
            swap(nums[i], nums[j]);
        }
        swap(nums[l], nums[j]);
        quickSort(nums, l, j - 1), quickSort(nums, j + 1, u);
    }
    
    void swap(int& a, int&b) {
        int tmp(move(a));
        a = move(b);
        b = move(tmp);
    }
public:
    vector<int> sortArray(vector<int>& nums) {
        quickSort(nums, 0, nums.size() - 1);
        return nums;
    }
};

File: Problems/912. Sort an Array/Solution2.js
// Merge Sort (in-place)
var sortArray = function(nums) {

    const merge = function(left, mid, right) {
        let i = left, j = mid + 1;
        while (i <= mid && j <= right) {
            if (nums[i] <= nums[j]) {
                ++i;
            } else {
                let tmp = nums[j];
                for (let k = j; k > i; --k) {
                    nums[k] = nums[k-1];
                }
                nums[i] = tmp;
                i ++;
                j ++;
                mid ++;
            }
        }
    }
    
    const mergeSort = function(left, right) {
        if (left >= right) return;
        
        let mid = left + right >> 1;
        mergeSort(left, mid);
        mergeSort(mid + 1, right);
        merge(left, mid, right);
    }
    
    mergeSort(0, nums.length - 1);
    return nums;
};